### PR TITLE
Define staking events

### LINKED ISSUE #3797: staking: inventorize useful staking events to index
The staking component has minimal event indexing (delegate and undelegate as of now).

Similarly to #3645 we should enumerate useful events to index (state transitions and such) and implement.

### LINKED ISSUE #4336: staking: emit ABCI events for validator state transitions
**Is your feature request related to a problem? Please describe.**
Currently, the most authoritative view of the active validator set is obtained by running `pcli query validator list --show-inactive`. We have partial, but crucially flawed, metrics related to validator states that should be removed.

**Describe the solution you'd like**
When a validator changes state, e.g. from Defined -> Active, or Active -> Jailed, or Active -> Tombstoned, an ABCI event should be emitted. This would allow external tooling like block explorers to consume these events.

**Describe alternatives you've considered**
We've tried to shore up the emitted metrics, but decided these aren't the right way to represent the validator state info. 

**Additional context**
See related historical discussion in https://github.com/penumbra-zone/penumbra/issues/3746, and recent PR in  EventTombstoneValidator - https://github.com/penumbra-zone/penumbra/pull/4277.

### PR DESCRIPTION
**Pull request by plaidfinch**
## Describe your changes

Resolves #4336, emitting events for validator definitions and state transitions. Included events:

- validator state change
- validator voting power change
- validator bonding state change
- validator rate data change
- validator definition upload
- validator missed block
- slashing penalty applied
- delegate
- undelegate
- funding stream rewards

All of these are proto-based events, and emitted both at genesis, when relevant, and when changed later.

Emitting events **at** genesis is technically impossible because the ABCI interface does not accept events as a response to `InitChain`, so instead genesis events are emitted at the start of the first block.

## Issue ticket number and link

Resolves #4336. Closes #3797. Closes #[4512](https://github.com/penumbra-zone/penumbra/issues/4512).

## Checklist before requesting a review

- [X] If this code contains consensus-breaking changes, I have added the "consensus-breaking" label. Otherwise, I declare my belief that there are not consensus-breaking changes, for the following reason:

  > This adds new events, but does not affect consensus.

### CODE REVIEW

File: proto/penumbra/penumbra/core/component/stake/v1/stake.proto
Comment (cratelyn) [inline proto/penumbra/penumbra/core/component/stake/v1/stake.proto:372]: :bike: :house: :paintbrush: with apologies, a bicycle shed

```suggestion
message EventValidatorRateDataChange {
```

should this also have the `EventValidator`, since it also holds an identity key and relates to the broader lifecycle of a validator? :woman_shrugging: i don't feel strongly about it, but it feels grouped with the rest of these.

File: crates/core/component/stake/src/component/validator_handler/validator_manager.rs
Comment (cratelyn) [inline crates/core/component/stake/src/component/validator_handler/validator_manager.rs:76]: thanks for doing this sort of thing while you are in the neighborhood. scout's rule :fleur_de_lis:

**Review by cratelyn — COMMENTED** [review]
this looks like it is coming together well! pardon the drive-by comments on a draft, i have one question to ask before these protobuf definitions ossify :angel:

File: crates/core/component/stake/src/component/action_handler/undelegate.rs
Comment (zbuc) [inline crates/core/component/stake/src/component/action_handler/undelegate.rs:88]: should we roll this into `push_undelegation`?

**Comment by zbuc** [comment]
This LGTM, I made a couple code organization changes and tested it and confirmed the deferred events work properly.

I have not tested it between upgrades however.

**Comment by hdevalence** [comment]
Is anything blocking merging this?

**Comment by zbuc** [comment]
> Is anything blocking merging this?

Only review

File: crates/core/app/src/app/mod.rs
Comment (hdevalence) [inline crates/core/app/src/app/mod.rs:167]: On reflection I have more concerns that this may be the wrong approach and I think we probably shouldn't do it.

**Review by hdevalence — CHANGES_REQUESTED** [review]
Concerns about data model (will elaborate inline)

File: crates/core/app/src/app/mod.rs
Comment (zbuc) [reply crates/core/app/src/app/mod.rs:167]: Is there another way we should handle it?

Comment (hdevalence) [reply crates/core/app/src/app/mod.rs:167]: @zbuc sorry, i got distracted doing the full writeup and was delayed. filled in the comment now

Comment (zbuc) [reply crates/core/app/src/app/mod.rs:167]: So the idea is that we remove this specially cased event emission and then let the indexers use genesis JSON to bootstrap initial state? Sounds fine to me.

Comment (zbuc) [reply crates/core/app/src/app/mod.rs:167]: @hdevalence indexers would also need to handle the genesis data checkpointed during upgrades.

**Review by aubrika — APPROVED** [review]
LGTM pending requested changes

**Review by zbuc — APPROVED** [review]

**Review by hdevalence — APPROVED** [review]

**Pull request merged**
